packages/catalog-model/examples/apis/swapi-graphql.yaml

```diff
@@ -3,6 +3,8 @@ kind: API
 metadata:
   name: starwars-graphql
   description: SWAPI GraphQL Schema
+  annotations:
+    graphql/endpoint: 'https://swapi-graphql.netlify.app/.netlify/functions/index'
```

Comment:
Annotation prefixes typically the domain TLD on them so this probably should be "graphql.org".

However, I'm not sure how universally useful this will be since it will only work for non authenticated endpoints. It feels a little funny adding an annotation to handle something that OpenAPI handles in its standard spec. I've also, never been a fan of using the graphiql interface here for docs since it's clearly not optimized for docs display. It was intended as a stop gap for something better, but unfortunately I have yet to come across anything that can easily work without introspection or doesn't generate static HTML.

I'm wondering if this feature could be added in a way that doesn't introduce a well known annotation that possible makes assumptions on the actual widget being used for docs? The only think of it moving this to the API entity spec or using a modal dialog to configure it at runtime. Neither of which are nice options.

Reply:
I agree it's not an ideal solution but I was trying to work with what Backstage already has. As for graphql being in the openapi specs, yes it is possible to execute graphql from swagger but for graphql queries, I've found graphiql to be easier to use and more informative. If the point of the Backstage Catalog is service discovery, then wouldn't graphiql be better? Since it auto completes as you type your query and has the doc explorer side tab that pops out (which again, I agree isn't ideal but better than nothing).

As for using an annotation, would perhaps using `graphql.org/graphiqlEndpoint` be enough of a distinction for users? Or is that being too specific and tied into graphiql?

What does moving it into the API entity spec entail?

Would using a modal dialog to configure at runtime mean that the user has to input the graphql endpoint every time the user goes to the api definition tab?

Considering all this, is there a good path forward for me to execute?

Reply:
@DanCNo - We should definitely open this up for others opinions, but here's my take ...

I think the general idea around this makes sense, but to clarify I just think the annotation approach will only be useful in a small set of uses cases where no auth is required and the endpoint is accessible to the user. The problem with the annotation approach is they are hard to extend with additional use cases without causing breaking changes. Therefore introducing one should be carefully thought out imo.

Also, when using the approach your suggesting there is actually no need to even use the API entity `spec.definition` because `graphiql` could likely just introspect the endpoint. Can I ask how many `GraphQL` endpoints your looking to define this way?

A couple of out of the box options are:

1 - Create a custom API Docs widget (see [readme](https://github.com/backstage/backstage/tree/master/plugins/api-docs#custom-api-renderings)) within your own project. This would allow you to use your particular use case and possibly iterate on it internally while we discuss a more future proof path forward.

2 - Extend the `@backstage/plugin-graphiql` plugin to contain tabs for each API you'd like to use this way.

I've been patiently waiting for `graphiql` v2 to ship which should be more modular and hopefully make it possible to compose a API docs tailored version together :)

Another possible path forward here could be to make this configurable with a modal dialog similar to how the OpenAPI one works. That said, the trick is being able to default values from somewhere. I think this is where each org could differ in how and what is required to use `graphiql` + introspection. I'm not convinced an annotation is flexible enough for this. 

Maybe what is actually needed is some sort of `GraphQLAPIDocsCredentialProvider` interface. This interface could take the entity name and then return a promise to the introspection route (or the entire schema). That way the backstage integrator provides the details (or pipes it through the backend proxy). Sorry just thinking out load, nothing concrete here.

Reply:
Sorry for the late response. Had to discuss with my team and another team. 
I don't know the exact number but we probably have over 100 graphql endpoints. 
So we are going to try to go with option no.1 and create a custom API docs widget and just have it override the spec type graphql.

Reply:
@DanCNo. I definitely think this sort of feature should be available, it's just figuring out the best way to do this in a flexible way. It's unfortunate there is no higher level spec for GraphQL that describes the API and not just the schema (SDL). Both OpenAPI and AysncAPI have optional server and security scheme metadata that can be used to help with this.

Thinking about this a bit more, I think what's missing is essentially that. A generalized way to provide the server and auth information for GraphQL. This could be something that belongs in the entity spec either directly or some envelope wrapper for graphql schemas (SDL). I wish there was something more formal around this though.

I'm wondering if maybe moving this concept into an RFC might be a good way to hash out something more formal?

Reply:
A related discussions here #5652.

Reply:
I am thinking that much of this PR makes sense and is aligned the discussion in #5652. The discussion suggests that we use custom renderers for the GraphQL API type. Perhaps we can enable that as a pattern but skip the well-known annotation for now, and leave that part out of this PR? just adding enough things to make it possible to implement. That way we make it possible for people to experiment a bit without committing too much to a particular approach in the catalog model.

Reply:
I agree. My main concern was around the inflexibility of the annotation approach. Having a code based approach with a limited scoped API would be a good short term solution.